src/html-transform.ts

```diff
@@ -0,0 +1,256 @@
+/**
+ * @license
+ * Copyright (c) 2018 The Polymer Project Authors. All rights reserved.
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ * The complete set of authors may be found at
+ * http://polymer.github.io/AUTHORS.txt
+ * The complete set of contributors may be found at
+ * http://polymer.github.io/CONTRIBUTORS.txt
+ * Code distributed by Google as part of the polymer project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+import * as dom5 from 'dom5';
+import * as parse5 from 'parse5';
+
+import {jsTransform, JsTransformOptions} from './js-transform'
+
+const p = dom5.predicates;
+
+const isJsScriptNode = p.AND(
+    p.hasTagName('script'),
+    p.OR(
+        p.NOT(p.hasAttr('type')),
+        p.hasAttrValue('type', 'text/javascript'),
+        p.hasAttrValue('type', 'application/javascript'),
+        p.hasAttrValue('type', 'module')));
+
+/**
+ * Options for htmlTransform.
+ */
+export interface HtmlTransformOptions {
+  // Transformations to apply to JavaScript within the HTML document.
+  js?: JsTransformOptions;
+
+  // Whether this is the top-level "entry point" HTML document.
+  isEntryPoint?: boolean;
+
+  // If js.transformEsModulesToAmd and isEntryPoint are both true (in which case
+  // this parameter is required), a URL for the Require.Js library, for which a
+  // script tag will be injected.
+  requireJsUrl?: string;
+}
+
+/**
+ * Transform some HTML according to the given options.
+ */
+export function htmlTransform(
+    html: string, options: HtmlTransformOptions): string {
+  if (options.js && options.js.transformEsModulesToAmd &&
+      options.isEntryPoint && !options.requireJsUrl) {
+    throw new Error(
+        'requireJsUrl is required when ' +
+        'js.transformEsModulesToAmd and isEntryPoint are true.');
+  }
+  if (options.js && options.js.moduleResolution === 'node' &&
+      !options.js.filePath) {
+    throw new Error('Cannot perform node module resolution without filePath.');
+  }
+
+  const document = parse5.parse(html);
+  const allScripts = dom5.queryAll(document, isJsScriptNode);
+
+  let wctScript;
+  const moduleScripts = [];
+  const otherScripts = [];
+  let shouldTransformEsModuleToAmd =
+      options.js && options.js.transformEsModulesToAmd;
+
+  for (const script of allScripts) {
```

Comment:
Are the two passes needed? It seems like given that we want to handle defer scripts correctly eventually, we could just loop through all the scripts, keep the state needed to enforce the ordering, and dispatch to helper methods based on type. Then we don't have 3 loops, which might be easier to read too.

Reply:
Done.